### ISSUE TITLE
Execution of the compiled query with lazy loading throws an exception

### ISSUE THREAD

**Issue by iveksp** [dotnet/efcore#14423]
Program throws a invalid operation exception when we use lazy loading and execute a compiled query in the same database context. It's only throws when is accessed to navigation property of the entity object which is passed as a parameter in the compiled query.

**Exception message:** 
A second operation started on this context before a previous operation completed. This is usually caused by different threads using the same instance of DbContext, however instance members are not guaranteed to be thread safe. This could also be caused by a nested query being evaluated on the client, if this is the case rewrite the query avoiding nested invocations.

**Stack trace:**
   ```
at Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.Load[TSource](IQueryable`1 source)
   at Microsoft.EntityFrameworkCore.Internal.EntityFinder`1.Load(INavigation navigation, InternalEntityEntry entry)
   at Microsoft.EntityFrameworkCore.Internal.LazyLoader.Load(Object entity, String navigationName)
   at Microsoft.EntityFrameworkCore.Infrastructure.LazyLoaderExtensions.Load[TRelated](ILazyLoader loader, Object entity, TRelated& navigationField, String navigationName)
   at ConsoleApp.ExistingDb.Blog.get_Post() in C:\Users\Ivan\Desktop\EntityFramework.Docs.git\trunk\samples\core\GetStarted\FullNet\ConsoleApp\Blog.cs:line 33
   at lambda_method(Closure , Post )
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at lambda_method(Closure )
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ResultEnumerable`1.GetEnumerator()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass15_1`1.<CompileQueryCore>b__0(QueryContext qc)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryBase`2.ExecuteCore(TContext context, CancellationToken cancellationToken, Object[] parameters)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryBase`2.ExecuteCore(TContext context, Object[] parameters)
   at ConsoleApp.Program.Main(String[] args) in C:\Users\Ivan\Desktop\EntityFramework.Docs.git\trunk\samples\core\GetStarted\FullNet\ConsoleApp\Program.cs:line 15
```

### Steps to reproduce
```
static void Main(string[] args)
{
     using (BloggingContext ctx = new BloggingContext())
     {
           Blog blog = ctx.Blog.FirstOrDefault();
           Post post = CompiledQueryTest(ctx, blog);
     }
 }
```

        public static Func<BloggingContext, Blog, Post> CompiledQueryTest = 
            EF.CompileQuery<BloggingContext, Blog, Post>((ctx, blog) => ctx.Post.FirstOrDefault(c => blog.Post.Any() && c == blog.Post.FirstOrDefault()));`

In the following cases this compiled query runs normally:
 ``` 
using (BloggingContext ctx = new BloggingContext())
{
     Blog blog = ctx.Blog.Include(c => c.Post).FirstOrDefault();
     Post post = CompiledQueryTest(ctx, blog);
 }
```

```
using (BloggingContext ctx = new BloggingContext())
{
     Blog blog = ctx.Blog.Include(c => c.Post).FirstOrDefault();
     using (BloggingContext ctx1 = new BloggingContext())
     {
           Post post = CompiledQueryTest(ctx1, blog);
     }
}
```

### Further technical details
EF Core version: 2.2.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 
IDE: Visual Studio 2017 15.9.5

**Comment by wensaint** [comment]
Has the 3.0 version fixed the problem?

**Comment by ajcvickers** [comment]
@wensaint No, the issue was moved to the Backlog, which means that a fix didn't make it into the 3.0 release.

**Comment by smitpatel** [comment]
Error in 5.0 
> Translation of 'EF.Property<Nullable<int>>(__blog.Posts
    .FirstOrDefault(), "Id")' failed.

Blocked on #19209
